[ANALYZER-3289] Issue of reverse gradient color in GeoMap when using …

mchen-len-son opened the pull request:
…csv as a datasource

- fixed geo points gradient coloring
- added unit test

mchen-len-son commented:
@rfellows @bennychow @carlosrusso please review

wingman-pentaho commented:
### Build Completed

:white_check_mark: This pull request has passed all validations.

#### Build Commands

```sh
ant -Dpentaho.resolve.repo=http://ivy-nexus.pentaho.org/content/groups/approved -Dtestreports.dir=bin/reports/unit-test -f build.xml clean-all resolve resolve-js jacoco && ant -Dpentaho.resolve.repo=http://ivy-nexus.pentaho.org/content/groups/approved -Dtestreports.dir=bin/reports/integration-test -f build.xml resolve resolve-js dist jacoco-integration test-js checkstyle
```

#### Cleanup Commands
```sh
rm -r ~/.ivy2/local || echo no publish local to remove
```

#### Changed files

```console
  package-res/resources/web/vizapi/VizController.js
  test-js/unit/vizapi/VizControllerSpec.js
```

#### Unit test coverage change

These statistics help you identify how your changes have affected the coverage of the following files. If a file is not in this list, then its coverage was not affected by your changes. To get some help interpreting these metrics, please refer to [Jacoco's documentation](http://eclemma.org/jacoco/trunk/doc/counters.html).

###### org.pentaho.common.ui.metadata.model.impl.Query
 * Complexity Change:  + 5.88%
 * Instruction Change:  + 5.33%
 * Line Change:  + 6.67%
 * Method Change:  + 5.88%

carlosrusso commented on package-res/resources/web/vizapi/VizController.js:
```diff
@@ -717,28 +717,28 @@ pentaho.VizController.getRgbGradientFromMultiColorHex = function(value, min, max
-    var end = Math.ceil(((value-min)/range) * steps);
+    var inRange = ((value-min)/range);
+    var startStep = Math.floor(inRange * steps);
+    var endStep = Math.ceil(inRange * steps);
```
Scope of variables: the variables `inRange`, `startStep` and `endStep` are used outside of this `if` block, and as such they should be declared outside.

mchen-len-son commented:
@carlosrusso 
updated the PR: moved out startStep and endStep.
inRange is actually not used outside the block it's declared in.
Thanks.

wingman-pentaho commented:
### Build Completed

:white_check_mark: This pull request has passed all validations.

#### Build Commands

```sh
ant -Dpentaho.resolve.repo=http://ivy-nexus.pentaho.org/content/groups/approved -Dtestreports.dir=bin/reports/unit-test -f build.xml clean-all resolve resolve-js jacoco && ant -Dpentaho.resolve.repo=http://ivy-nexus.pentaho.org/content/groups/approved -Dtestreports.dir=bin/reports/integration-test -f build.xml resolve resolve-js dist jacoco-integration test-js checkstyle
```

#### Cleanup Commands
```sh
rm -r ~/.ivy2/local || echo no publish local to remove
```

#### Changed files

```console
  package-res/resources/web/vizapi/VizController.js
  test-js/unit/vizapi/VizControllerSpec.js
```

#### Unit test coverage change

These statistics help you identify how your changes have affected the coverage of the following files. If a file is not in this list, then its coverage was not affected by your changes. To get some help interpreting these metrics, please refer to [Jacoco's documentation](http://eclemma.org/jacoco/trunk/doc/counters.html).

###### org.pentaho.common.ui.metadata.model.impl.Query
 * Complexity Change:  + 5.88%
 * Instruction Change:  + 5.33%
 * Line Change:  + 6.67%
 * Method Change:  + 5.88%

carlosrusso commented on package-res/resources/web/vizapi/VizController.js:
```diff
@@ -716,29 +716,31 @@ pentaho.VizController.getRgbGradientFromMultiColorHex = function(value, min, max
+    var inRange = ((value-min)/range);
+    startStep = Math.floor(inRange * steps);
+    endStep = Math.ceil(inRange * steps);
   }
```
Are you sure this is right? The difference between  `startStep` and `endStep` is either 0 or 1, which means `color1` and `color2` will either be the same color or adjacent.

carlosrusso replied:
Nevermind the comment above, I failed to follow the entire logic. `startStep` and `endStep` identify the indexes in the color palette that need to be scaled.

carlosrusso replied:
Since both `startStep` and `endStep` are indexes of the `colors` array, we could rename them `startColorIndex` and `endColorIndex` so that future readers of this code endure fewer pains.

carlosrusso commented on package-res/resources/web/vizapi/VizController.js:
```diff
@@ -716,29 +716,31 @@ pentaho.VizController.getRgbGradientFromMultiColorHex = function(value, min, max
-    inRange = 1;
+  var inStepRange;
+  if(startStep == endStep){
+    inStepRange = 0;
```
Instead of setting `inStepRange = 0` and doing useless math further below, you just have a: `return pentaho.VizController.getRrbColor(color1[0], color1[1], color1[2]);` here

carlosrusso commented on package-res/resources/web/vizapi/VizController.js:
```diff
@@ -716,29 +716,31 @@ pentaho.VizController.getRgbGradientFromMultiColorHex = function(value, min, max
-    inRange = (value-rangeMin)/(rangeMax-rangeMin);
+    var stepRangeMin = ((startStep / steps) * range) + min;
+    var stepRangeMax = ((endStep / steps) * range) + min;
+    inStepRange = (value-stepRangeMin)/(stepRangeMax-stepRangeMin);
```
Personally, I find the names `inRange` and `inStepRange` is a bit obscure.
What about renaming the variable `inStepRange` to `scalingFactor` or something more suggestive?

mchen-len-son commented:
@carlosrusso I updated the PR.
As for variables naming, I've updated them to a bit closer to their meaning,
but `scalingFactor` is about different things though.

wingman-pentaho commented:
### Build Completed

:x: This pull request has errors. They will need to be addressed before it can be accepted. See below for more details. Some links are also available below for further assistance in addressing those issues.

#### Build Commands

```sh
ant -Dpentaho.resolve.repo=http://ivy-nexus.pentaho.org/content/groups/approved -Dtestreports.dir=bin/reports/unit-test -f build.xml clean-all resolve resolve-js jacoco && ant -Dpentaho.resolve.repo=http://ivy-nexus.pentaho.org/content/groups/approved -Dtestreports.dir=bin/reports/integration-test -f build.xml resolve resolve-js dist jacoco-integration test-js checkstyle
```

#### Cleanup Commands
```sh
rm -r ~/.ivy2/local || echo no publish local to remove
```

#### Changed files

```console
  package-res/resources/web/vizapi/VizController.js
  test-js/unit/vizapi/VizControllerSpec.js
```

Newly Broken Tests:
======================
###### org.pentaho.common.ui.metadata.service.MetadataService2IT.testJsonQueryToJson1:

:red_circle: `junit.framework.AssertionFailedError`

```java
junit.framework.AssertionFailedError: wrong value
	at org.pentaho.common.ui.metadata.service.MetadataService2IT.testJsonQueryToJson1(MetadataService2IT.java:731)

```
###### org.pentaho.common.ui.metadata.service.MetadataService2IT.testJsonQueryToJson2:

:red_circle: `junit.framework.AssertionFailedError`

```java
junit.framework.AssertionFailedError: wrong value
	at org.pentaho.common.ui.metadata.service.MetadataService2IT.testJsonQueryToJson2(MetadataService2IT.java:750)

```
###### org.pentaho.common.ui.metadata.service.MetadataService2IT.testXmlQueryToJson1:

:red_circle: `junit.framework.AssertionFailedError`

```java
junit.framework.AssertionFailedError: wrong value
	at org.pentaho.common.ui.metadata.service.MetadataService2IT.testXmlQueryToJson1(MetadataService2IT.java:711)

```

#### Unit test coverage change

These statistics help you identify how your changes have affected the coverage of the following files. If a file is not in this list, then its coverage was not affected by your changes. To get some help interpreting these metrics, please refer to [Jacoco's documentation](http://eclemma.org/jacoco/trunk/doc/counters.html).

###### org.pentaho.common.ui.metadata.model.impl.Query
 * Complexity Change:  + 11.76%
 * Instruction Change:  + 10.67%
 * Line Change:  + 13.33%
 * Method Change:  + 11.76%

carlosrusso commented on package-res/resources/web/vizapi/VizController.js:
```diff
@@ -716,29 +716,31 @@ pentaho.VizController.getRgbGradientFromMultiColorHex = function(value, min, max
-    var start = Math.floor(((value-min)/range) * steps);
-    var end = Math.ceil(((value-min)/range) * steps);
+    var positionInRange = ((value-min)/range);
+    startColorIndex = Math.floor(positionInRange * steps);
```
what about having: `positionInRange = ((value-min)/range)*steps`? This would economize one multiplication
